File: solutions/fable-1/fable/ruby/brackets.rb
PAIRS = { ')' => '(', ']' => '[', '}' => '{' }.freeze
OPENERS = PAIRS.values.freeze

def balanced?(line)
  stack = []
  line.each_char do |c|
    if OPENERS.include?(c)
      stack.push(c)
    elsif PAIRS.key?(c)
      return false if stack.pop != PAIRS[c]
    else
      return false
    end
  end
  stack.empty?
end

out = []
$stdin.each_line do |raw|
  line = raw.chomp.delete("\r")
  out << (balanced?(line) ? 'yes' : 'no')
end
puts out.join("\n") unless out.empty?

File: solutions/fable-1/fable/perl/brackets.pl
use strict;
use warnings;

my %pair = (')' => '(', ']' => '[', '}' => '{');

while (my $line = <STDIN>) {
    $line =~ s/\r?\n\z//;
    my @stack;
    my $ok = 1;
    for my $c (split //, $line) {
        if ($c eq '(' || $c eq '[' || $c eq '{') {
            push @stack, $c;
        }
        elsif (exists $pair{$c}) {
            if (!@stack || pop(@stack) ne $pair{$c}) {
                $ok = 0;
                last;
            }
        }
        else {
            $ok = 0;
            last;
        }
    }
    $ok = 0 if @stack;
    print $ok ? "yes\n" : "no\n";
}
